Add header files to support independent compilation of volesti headers

Neel-Shah-29 opened the pull request:
This PR fixes #184 
According to the issue  , The `cmath`  header file was not included in `boundary_cdhr_walk.hpp`  , which was giving errors of pow , sqrt functions. So i have made the necessary changes. Kindly suggest if any other changes needs to be done.

Commit:
header file of math included in boundary_cdhr_walk.hpp

vissarion reviewed (commented):
Thanks for this PR. 

Did you check that now the header files compile independently? Adding toy examples that include headers files of the library and compile independently would be helpful.

Neel-Shah-29 commented:
> Thanks for this PR.
> 
> Did you check that now the header files compile independently? Adding toy examples that include headers files of the library and compile independently would be helpful.

Yes some examples are running correctly but i got an error while building `Try1.cpp` which is mentioned in the issue #184 

![image](https://user-images.githubusercontent.com/84740927/156027573-8f1219fa-0450-4542-b56d-e33ad6f4c62f.png)

This error will be resolved my including `cmath` in `sphere.hpp` and including `bits/stdc++.h` in `boundary_cdhr_walk.hpp`.
This solution worked fine for me!
![image](https://user-images.githubusercontent.com/84740927/156029302-a3fb34ea-bb00-471b-8039-6dbe02c7fbed.png)

All the tests and build ran successfully after making that change.Now the examples compile independently without explicitly including `cmath` in the example program.

So should i make the changes and put a commit for the same ?

vissarion reviewed (changes requested):
Thanks! Some thoughts here.

* I would suggest not including `bits/stdc++.h` since it is not standard and will include a lot of unnecessary headers. 

* Also your solution currently partially solves the issue; see what happens if one tries to include `boundary_rdhr_walk.hpp` instead of `random_walks.hpp`. Similar, issues are expected with some other walks too e.g. `boltzmann_hmc_walk.hpp`.

Commit:
Changes requested were made

Neel-Shah-29 commented:
> Thanks! Some thoughts here.
> 
> * I would suggest not including `bits/stdc++.h` since it is not standard and will include a lot of unnecessary headers.
> * Also your solution currently partially solves the issue; see what happens if one tries to include `boundary_rdhr_walk.hpp` instead of `random_walks.hpp`. Similar, issues are expected with some other walks too e.g. `boltzmann_hmc_walk.hpp`.

* I have removed `bits/stdc++.h` and instead added `utility` header file which was required for using `pair` in `boundary_cdhr_walk.hpp` and `boundary_rdhr_walk.hpp`
* I was getting errors of `list` in  `boltzmann_hmc_walk.hpp`  on including it in `Try1.cpp` so i included  `list` in that file.
* I added `cmath` in `sphere.hpp` as it was required according to above comment

Do i need to check importing all walks files in example and try compiling them independent of header files or this much changes are enough? 

Kindly suggest me if any changes are to be done. Thanks!

vissarion commented on include/random_walks/boltzmann_hmc_walk.hpp:
```diff
@@ -213,4 +213,4 @@ struct BoltzmannHMCWalk {
 };
 
-#endif //VOLESTI_BOLTZMANN_HMC_WALK_HPP
+#endif //VOLESTI_BOLTZMANN_HMC_WALK_HPP
```
please do not remove the new lines at the end of files

vissarion commented:
> Kindly suggest me if any changes are to be done. Thanks!

Could you please check all files in `random_walks` one by one if they compile alone and add the necessary header files to them?

Neel-Shah-29 commented:
> > Kindly suggest me if any changes are to be done. Thanks!
> 
> Could you please check all files in `random_walks` one by one if they compile alone and add the necessary header files to them?

Yes sure i will check for all of them. 

I have tried compiling `exponential_hamiltonian_monte_carlo_exact_walk.hpp` than i am getting the following error.
Similar error is coming while compiling  `gaussian_hamiltonian_monte_carlo_exact_walk.hpp`  also.

![image](https://user-images.githubusercontent.com/84740927/156430694-ab124478-09ed-4f49-bc85-795460a90b78.png)

It says `compute_diameter` is not included in scope ,Can you please tell the header file that needs to be included in both of these files to resolve this error.

Neel-Shah-29 commented:
> > > Kindly suggest me if any changes are to be done. Thanks!
> > 
> > 
> > Could you please check all files in `random_walks` one by one if they compile alone and add the necessary header files to them?
> 
> Yes sure i will check for all of them.
> 
> I have tried compiling `exponential_hamiltonian_monte_carlo_exact_walk.hpp` than i am getting the following error. Similar error is coming while compiling `gaussian_hamiltonian_monte_carlo_exact_walk.hpp` also.
> 
> ![image](https://user-images.githubusercontent.com/84740927/156430694-ab124478-09ed-4f49-bc85-795460a90b78.png)
> 
> It says `compute_diameter` is not included in scope ,Can you please tell the header file that needs to be included in both of these files to resolve this error.

This issue is resolved by including `uniform_billiard_walk.hpp` in both the files as `compute_diameter ` is declared in over there.

vissarion commented:
> This issue is resolved by including `uniform_billiard_walk.hpp` in both the files as `compute_diameter ` is declared in over there.

OK, please add a comment on the code on why you included that header file. What about all other header files? Are they OK?

Commit:
All the header files under random_walks compiles independently

vaithak commented:
Hi @vissarion, I think this check should be automated so that this is ensured in every PR. 
I find this stackoverflow answer helpful: https://stackoverflow.com/a/57900525/7317290. @Neel-shah-29 can you check this out?

Neel-Shah-29 commented:
> Hi @vissarion, I think this check should be automated so that this is ensured in every PR. I find this stackoverflow answer helpful: https://stackoverflow.com/a/57900525/7317290. @Neel-Shah-29 can you check this out?

Yes sure i will look into this answer.

> > This issue is resolved by including `uniform_billiard_walk.hpp` in both the files as `compute_diameter ` is declared in over there.
> 
> OK, please add a comment on the code on why you included that header file. What about all other header files? Are they OK?

Hello @vissarion I have added all the necessary header files in all the files of  `random_walks`  and checked compiling all of them independently.

All the files under `random_walks` compiles independently now according to me. Please review the PR and suggest if any necessary changes needs to be made. Thanks!

vissarion commented:
Hi @Neel-Shah-29 thanks. Would you like to try the automated procedure proposed by @vaithak ?

Neel-Shah-29 commented:
> Hi @Neel-Shah-29 thanks. Would you like to try the automated procedure proposed by @vaithak ?

Yes i can try that approach as well, but can this PR be merged first and then i try that approach, as issue #184 can be resolved by this PR. I will submit another PR for that approach,is that ok?

vaithak commented:
Hi @Neel-Shah-29, I would prefer that the automated check be done in this PR only, reason being that whenever a bug is solved, the corresponding PR should add some form of check/tests ensuring that this bug won't happen again in the future. This check will ensure that, also this will be make it easy for the reviewers to verify that the problem is actually solved for all the header files.

Commit:
Custom test added for compiling header files independently

Neel-Shah-29 commented:
Hello @vaithak i have pushed the code after adding the custom test kindly check whether its ok or do i need to make any changes. After running `make chkdeps` i am getting the following output
![image](https://user-images.githubusercontent.com/84740927/156933812-31a314bf-971d-4032-aff4-f578510fb960.png)

vaithak commented:
Hi @Neel-Shah-29, if you see your log message no target is made for any of the header files. You would have to change the glob expression for searching files. 
I think this should work:
```
file(GLOB_RECURSE HDR_ROOT "../include/*.h" "../include/*.hpp")
```

Commit:
Custom test added for compiling header files independently

Neel-Shah-29 commented:
Hello @vaithak this solution worked great , i am getting target to all header files successfully as seen in below screenshot.
![image](https://user-images.githubusercontent.com/84740927/157002945-aab4b7c1-d7af-4121-ad99-45317b1dd792.png)
but i got an error regarding some target files already exists in `/home/neel/volesti/test` in my local computer but otherwise everything works fine!
![image](https://user-images.githubusercontent.com/84740927/157003835-1c09f0f6-9c64-482a-984f-9a36490b7dff.png)
I am pushing a commit after doing the changes if any further corrections needs to be done than kindly report it to me

vaithak commented:
@Neel-Shah-29, I think that is because of only using filenames without extensions for creating the custom targets, you can also include the extension for naming the target. You will have to refer to cmake's documentation on how that can be done.

Neel-Shah-29 commented:
> @Neel-Shah-29, I think that is because of only using filenames without extensions for creating the custom targets, you can also include the extension for naming the target. You will have to refer to cmake's documentation on how that can be done.

Hello @vissarion @vaithak ,This problem was solved by replacing `NAME_WE` by `NAME` as follows:-
`get_filename_component(HDR_WE ${HDR} NAME)`
All the files are included with its extension

But i got an error regarding `Eigen/Dense: No such file or directory` . I tried to find in which file Eigen/Dense was declared but i didn't find it. Can you please help me resolve the following error .
![image](https://user-images.githubusercontent.com/84740927/157089539-8548df94-111b-44c4-8f91-697abd78a22b.png)

vaithak commented:
Hi @Neel-Shah-29. I suspected this would happen as a single header file can include other header files. 
A fix for this can be to provide the path to essential header files while compiling this custom target. 
This can be done by adding the following into the COMMAND argument of add_custom_target.
```
-I../external/_deps/eigen-src -I../external/_deps/boost-src -I../include
```

This may also require setting WORKING-DIRECTORY of add_custom_target as we are using relative paths with -I

Neel-Shah-29 commented:
Hi @vaithak The solution given by you worked fine for me , i have added the `WORKING_DIRECTORY` and  argument in `COMMAND` as mentioned above but i got a error for the header file `utils.h`. Can you suggest what can be its possible solution.

![image](https://user-images.githubusercontent.com/84740927/157200616-855387fb-9dbf-4ee2-8eab-0405e6e0b38f.png)

vaithak commented:
The issue is same here, you have to provide the include directories for these header files.

Commit:
Custom test added to circleci for compilation of header files indeoendently

vaithak commented on test/CMakeLists.txt:
```diff
@@ -343,6 +343,20 @@ else ()
   TARGET_LINK_LIBRARIES(logconcave_sampling_test ${LP_SOLVE} ${IFOPT} ${IFOPT_IPOPT} ${PTHREAD} ${GMP} ${MPSOLVE} ${FFTW3} ${MKL_LINK} coverage_config)
   TARGET_LINK_LIBRARIES(order_polytope ${LP_SOLVE} coverage_config)
-
+ 
```
I would suggest adding comments here regarding the purpose of adding this.
Also add `make chkdeps` in circleci tests also.

Neel-Shah-29 committed:
Update CMakeLists.txt

Neel-Shah-29 commented:
Hello @vaithak , I have made the necessary changes but since the `WORKING_DIRECTORY` is not set properly the circle ci tests is failing. It runs properly on my local computer but here it shows error, what can be the possible `WORKING_DIRECTORY`  for running circle ci test.

vaithak commented:
You can use CMAKE_SOURCE_DIR variable to set the absolute path of working directory.

vaithak commented on include/volume/volume_cooling_balls.hpp:
```diff
@@ -12,7 +12,7 @@
 #include <boost/math/distributions/students_t.hpp>
 #include <boost/math/special_functions/erf.hpp>
-
+#include <volume/sampling_policies.hpp>
```
use `""` instead of `<>` as non standard header.

vaithak commented on include/volume/rotating.hpp:
```diff
@@ -10,7 +10,8 @@
 
 #include <Eigen/Eigen>
-
+#include<boost/random.hpp>
```
for proper formatting, ensure space between `#include` and `<`

vaithak commented on include/sampling/sphere.hpp:
```diff
@@ -10,7 +10,7 @@
 #ifndef SAMPLERS_SPHERE_HPP
 #define SAMPLERS_SPHERE_HPP
-
+#include<cmath>
```
same here, spacing.

vaithak commented on include/random_walks/gaussian_helpers.hpp:
```diff
@@ -2,7 +2,8 @@
 #define EXP_CHORD_TOLERENCE 0.00000001
-
+#include<algorithm>
+#include<cmath>
```
spacing

vaithak commented on include/ode_solvers/generalized_leapfrog.hpp:
```diff
@@ -10,7 +10,10 @@
+#include<vector>
+#include<Eigen/Eigen>
+#include<utility>
+#include<iostream>
```
spacing

vaithak commented on include/ode_solvers/leapfrog.hpp:
```diff
@@ -10,6 +10,7 @@
 
 #ifndef ODE_SOLVERS_LEAPFROG_HPP
 #define ODE_SOLVERS_LEAPFROG_HPP
+#include "Eigen/Eigen"
```
everywhere in the code, we have used `<>` for eigen and boost, so I think we should keep it consistent.

vaithak commented on include/ode_solvers/randomized_midpoint.hpp:
```diff
@@ -12,7 +12,7 @@
 #ifndef ODE_SOLVERS_RANDOMIZED_MIDPOINT_HPP
 #define ODE_SOLVERS_RANDOMIZED_MIDPOINT_HPP
-
+#include "Eigen/Eigen"
```
same here.

Neel-Shah-29 commented:
Hey @vaithak I have made the above changes on my local computer , I will be pushing a commit in a while 
![image](https://user-images.githubusercontent.com/84740927/158157642-89ad0d92-1f23-47a7-b78f-a3fd74989fe6.png)
I am getting the above error that `variable_set.h` is not included , i am not able to find this header file in any of the sections in the codebase,What should i include to rectify this error?

vissarion commented:
It seems that this PR is stale for a long time. I marked it `stale` as a reminder and I will close it is it remains stale after 4 weeks.
